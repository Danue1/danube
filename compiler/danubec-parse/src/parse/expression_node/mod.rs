mod literal_expression_node;

use crate::State;

impl crate::Context {
    pub fn expression_node(&mut self) -> State {
        one_of!(self.expression_literal_node())
    }
}
