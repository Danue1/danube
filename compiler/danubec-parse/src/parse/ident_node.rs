use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn ident_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, IDENT_NODE);
        self.finish_node();
        State::Stop
    }
}
