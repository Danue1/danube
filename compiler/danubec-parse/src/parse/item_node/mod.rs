mod enum_item_node;
mod struct_item_node;

impl crate::Context {
    pub fn item_node(&mut self) -> crate::State {
        one_of!(self.struct_item_node(), self.enum_item_node())
    }
}
