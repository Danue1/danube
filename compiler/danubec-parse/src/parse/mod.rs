mod ident_node;
mod item_node;
mod path_node;
mod type_node;

use crate::context::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::context::Context {
    #[inline]
    pub fn ast(&mut self) {
        self.start_node(SyntaxKind::AST_NODE);
        self.skip_whitespace();
        while !self.is_empty() {
            if self.item_node() == State::Continue {
                self.unexpected_token();
            }
            self.skip_whitespace();
        }
        self.finish_node();
    }

    fn unexpected_token(&mut self) {
        self.start_node(SyntaxKind::ERROR);
        self.bump();
        self.finish_node();
    }
}
