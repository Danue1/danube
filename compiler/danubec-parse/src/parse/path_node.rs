use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn path_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, PATH_NODE, ident_node);

        while !self.is_eof() {
            if expect!(self, SyntaxKind::COLON) {
                if expect!(self, SyntaxKind::COLON) {
                    self.skip_whitespace();
                    self.ident_node();
                } else {
                    self.unexpected_token();
                }
            } else {
                break;
            }
        }

        self.finish_node();
        State::Stop
    }
}
