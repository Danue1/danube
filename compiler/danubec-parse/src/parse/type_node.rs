use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn type_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.unnamed_type_kind_node() == State::Continue
            && self.path_type_kind_node() == State::Continue
        {
            State::Continue
        } else {
            self.start_node_at(checkpoint, SyntaxKind::TYPE_NODE);
            self.finish_node();
            State::Stop
        }
    }

    // ()
    // (type1)
    // (type1, type2)
    pub fn unnamed_type_kind_node(&mut self) -> State {
        guard!(self, LEFT_PAREN, UNNAMED_TYPE_KIND_NODE);
        self.skip_whitespace();

        while !self.is_eof() {
            if expect!(self, SyntaxKind::RIGHT_PAREN) {
                break;
            }

            let checkpoint = self.checkpoint();
            if self.type_node() == State::Continue {
                self.unexpected_token();
            } else {
                self.start_node_at(checkpoint, SyntaxKind::UNNAMED_TYPE_KIND_NODE);
                self.finish_node();
            }
            self.skip_whitespace();
            if expect!(self, SyntaxKind::COMMA) {
                self.skip_whitespace();
            }
        }

        self.finish_node();
        State::Stop
    }

    // Foo
    // Foo::Bar
    pub fn path_type_kind_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.path_node() == State::Continue {
            return State::Continue;
        }

        self.start_node_at(checkpoint, SyntaxKind::PATH_TYPE_KIND_NODE);
        self.finish_node();
        State::Stop
    }
}
