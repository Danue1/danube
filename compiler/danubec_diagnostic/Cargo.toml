[package]
name = "danubec_diagnostic"
version = "0.1.0"
edition = "2024"

[dependencies]
