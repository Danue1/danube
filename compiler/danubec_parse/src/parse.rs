use crate::{
    event::Event,
    grammar::{Parser, krate},
    lower::lower_krate,
};
use danubec_ast::Krate;
use danubec_diagnostic::Diagnostic;
use danubec_lex::lex;
use danubec_syntax::{AstNode, GreenNodeBuilder, SyntaxKind, SyntaxNode};
use std::path::PathBuf;

pub fn parse_crate_main(path: &PathBuf) -> Result<Krate, Diagnostic> {
    parse_krate(path.join("src/main.dnb"))
}

pub fn parse_crate_lib(path: &PathBuf) -> Result<Krate, Diagnostic> {
    parse_krate(path.join("src/lib.dnb"))
}

pub fn parse_krate(parent: PathBuf) -> Result<Krate, Diagnostic> {
    let mut diagnostic = Diagnostic::new();
    let source = match read(&parent, &mut diagnostic) {
        Ok(source) => source,
        Err(()) => return Err(diagnostic),
    };
    let (node, mut diagnostic) = parse(&source, diagnostic);
    let krate = match danubec_syntax::Krate::cast(node) {
        Some(node) => lower_krate(node, &mut diagnostic).map_err(|_| diagnostic)?,
        None => {
            diagnostic.report(miette!("ICE: Expected krate node"));
            return Err(diagnostic);
        }
    };

    Ok(krate)
}

fn read(path: &PathBuf, diagnostic: &mut Diagnostic) -> Result<String, ()> {
    match std::fs::read_to_string(path) {
        Ok(source) => Ok(source),
        Err(error) => {
            diagnostic.report(miette!("Failed to read file {}: {}", path.display(), error));

            Err(())
        }
    }
}

pub(crate) fn parse(source: &str, diagnostic: Diagnostic) -> (SyntaxNode, Diagnostic) {
    let tokens = lex(source);
    let (events, diagnostic) = {
        let tokens: Vec<_> = tokens.iter().map(|&(kind, _)| kind).collect();
        let mut parser = Parser::new(&tokens, diagnostic);
        krate(&mut parser);
        parser.finish()
    };
    let node = build(&tokens, events);

    (node, diagnostic)
}

fn build(mut tokens: &[(SyntaxKind, &str)], events: Vec<Event>) -> SyntaxNode {
    let mut builder = GreenNodeBuilder::new();

    macro_rules! advance {
        () => {
            if let Some(&(kind, text)) = tokens.get(0) {
                builder.token(kind.into(), text);
                tokens = &tokens[1..];
            }
        };
    }

    macro_rules! trivia {
        () => {
            loop {
                match tokens.get(0) {
                    Some((kind, _)) if kind.at_trivia() => {
                        advance!();
                    }
                    _ => break,
                }
            }
        };
    }

    for event in events {
        match event {
            Event::Start(kind) => {
                builder.start_node(kind.into());
                trivia!();
            }
            Event::Token => {
                advance!();
                trivia!();
            }
            Event::End => builder.finish_node(),
            Event::Placeholder => {
                //
            }
        }
    }

    SyntaxNode::new_root(builder.finish())
}
