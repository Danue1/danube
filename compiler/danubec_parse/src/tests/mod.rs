use crate::parse;
use danubec_diagnostic::Diagnostic;

#[test]
fn top_level_attribute() {
    let source = r#"
#![a("value")]
#![a(key)]
#![a(key1, key2)]
#![a(key = "value")]
#![a::b(key = "value")]"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn attribute() {
    let source = r#"
#[a("value")]
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn zero_sized_struct() {
    let source = r#"
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn named_struct() {
    let source = r#"
struct Foo { }
struct Bar {
    a: i32,
    b: String,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn unnamed_struct() {
    let source = r#"
struct Foo();
struct Bar(());
struct Baz(i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn mixed_enum() {
    let source = r#"
enum Foo {}
enum Bar {
    A,
    B { a: i32, b: String },
    C(i32, String),
    D = 42,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn module() {
    let source = r#"
mod foo;
mod bar {
    struct Foo;
    enum Bar {
        A,
        B { a: i32, b: String },
        C(i32, String),
        D = 42,
    }
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn constant() {
    let source = r#"
const FOO: i32 = 42;
const BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn static_variable() {
    let source = r#"
static FOO: i32 = 42;
static BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn type_alias() {
    let source = r#"
type Foo = i32;
type Bar = (i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn function() {
    let source = r#"
fn foo();
fn bar() {}
fn baz(a: i32, b: String) -> (i32, String) {
    (a, b)
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn trait_definition() {
    let source = r#"
trait Foo { }
trait Bar {
    fn foo();
    const BAR: i32;
    type Baz;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn anonymous_implement() {
    let source = r#"
impl Foo { }
impl Foo {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn targeted_implement() {
    let source = r#"
impl Foo for Bar { }
impl Foo for Bar {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn use_definition() {
    let source = r#"
use {};
use {{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use *;
use foo;
use foo as bar;
use foo::{};
use foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};

use ::{};
use ::{{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use ::*;
use ::foo;
use ::foo as bar;
use ::foo::{};
use ::foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};
"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}
