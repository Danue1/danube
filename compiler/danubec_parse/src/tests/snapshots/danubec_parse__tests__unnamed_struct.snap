---
source: compiler/danubec_parse/src/tests/mod.rs
expression: events
---
KRATE_NODE@0..55
  NEW_LINE@0..1 "\n"
  STRUCT_DEFINITION_NODE@1..15
    STRUCT@1..7 "struct"
    WHITESPACE@7..8 " "
    IDENTIFIER_NODE@8..11
      IDENTIFIER_SEGMENT@8..11
        IDENTIFIER@8..11 "Foo"
    STRUCT_BODY_UNNAMED_NODE@11..15
      LEFT_PAREN@11..12 "("
      RIGHT_PAREN@12..13 ")"
      SEMICOLON@13..14 ";"
      NEW_LINE@14..15 "\n"
  STRUCT_DEFINITION_NODE@15..31
    STRUCT@15..21 "struct"
    WHITESPACE@21..22 " "
    IDENTIFIER_NODE@22..25
      IDENTIFIER_SEGMENT@22..25
        IDENTIFIER@22..25 "Bar"
    STRUCT_BODY_UNNAMED_NODE@25..31
      LEFT_PAREN@25..26 "("
      STRUCT_BODY_UNNAMED_FIELD_NODE@26..28
        TUPLE_TYPE_NODE@26..28
          LEFT_PAREN@26..27 "("
          RIGHT_PAREN@27..28 ")"
      RIGHT_PAREN@28..29 ")"
      SEMICOLON@29..30 ";"
      NEW_LINE@30..31 "\n"
  STRUCT_DEFINITION_NODE@31..55
    STRUCT@31..37 "struct"
    WHITESPACE@37..38 " "
    IDENTIFIER_NODE@38..41
      IDENTIFIER_SEGMENT@38..41
        IDENTIFIER@38..41 "Baz"
    STRUCT_BODY_UNNAMED_NODE@41..55
      LEFT_PAREN@41..42 "("
      STRUCT_BODY_UNNAMED_FIELD_NODE@42..45
        PATH_TYPE_NODE@42..45
          PATH_NODE@42..45
            PATH_SEGMENT_IDENTIFIER_NODE@42..45
              IDENTIFIER_NODE@42..45
                IDENTIFIER_SEGMENT@42..45
                  IDENTIFIER@42..45 "i32"
      COMMA@45..46 ","
      WHITESPACE@46..47 " "
      STRUCT_BODY_UNNAMED_FIELD_NODE@47..53
        PATH_TYPE_NODE@47..53
          PATH_NODE@47..53
            PATH_SEGMENT_IDENTIFIER_NODE@47..53
              IDENTIFIER_NODE@47..53
                IDENTIFIER_SEGMENT@47..53
                  IDENTIFIER@47..53 "String"
      RIGHT_PAREN@53..54 ")"
      SEMICOLON@54..55 ";"
